core/src/main/scala/akka/persistence/r2dbc/state/scaladsl/DurableStateDao.scala

```diff
@@ -202,61 +329,121 @@ private[r2dbc] class DurableStateDao(settings: R2dbcSettings, connectionFactory:
+
+    val entityType = PersistenceId.extractEntityType(state.persistenceId)
+
     val result = {
+      val additionalBindings = additionalColumns.get(entityType) match {
+        case None => Vector.empty[EvaluatedAdditionalColumnBindings]
+        case Some(columns) =>
+          val slice = persistenceExt.sliceForPersistenceId(state.persistenceId)
+          val upsert = AdditionalColumn.Upsert(state.persistenceId, entityType, slice, state.revision, value)
+          columns.map(c => EvaluatedAdditionalColumnBindings(c, c.bind(upsert)))
+      }
+
       if (state.revision == 1) {
-        val entityType = PersistenceId.extractEntityType(state.persistenceId)
         val slice = persistenceExt.sliceForPersistenceId(state.persistenceId)
 
-        r2dbcExecutor
-          .updateOne(s"insert [${state.persistenceId}]") { connection =>
-            val stmt = connection
-              .createStatement(insertStateSql)
-              .bind(0, slice)
-              .bind(1, entityType)
-              .bind(2, state.persistenceId)
-              .bind(3, state.revision)
-              .bind(4, state.serId)
-              .bind(5, state.serManifest)
-              .bind(6, state.payload.getOrElse(Array.emptyByteArray))
-            bindTags(stmt, 7)
-          }
-          .recoverWith { case _: R2dbcDataIntegrityViolationException =>
+        def insertStatement(connection: Connection): Statement = {
+          val stmt = connection
+            .createStatement(insertStateSql(entityType, additionalBindings))
+            .bind(getAndIncIndex(), slice)
+            .bind(getAndIncIndex(), entityType)
+            .bind(getAndIncIndex(), state.persistenceId)
+            .bind(getAndIncIndex(), state.revision)
+            .bind(getAndIncIndex(), state.serId)
+            .bind(getAndIncIndex(), state.serManifest)
+            .bind(getAndIncIndex(), state.payload.getOrElse(Array.emptyByteArray))
+          bindTags(stmt, getAndIncIndex())
+          bindAdditionalColumns(stmt, additionalBindings)
+        }
+
+        def recoverDataIntegrityViolation[A](f: Future[A]): Future[A] =
+          f.recoverWith { case _: R2dbcDataIntegrityViolationException =>
             Future.failed(
               new IllegalStateException(
                 s"Insert failed: durable state for persistence id [${state.persistenceId}] already exists"))
           }
+
+        changeHandlers.get(entityType) match {
+          case None =>
+            recoverDataIntegrityViolation(r2dbcExecutor.updateOne(s"insert [${state.persistenceId}]")(insertStatement))
+          case Some(handler) =>
+            r2dbcExecutor.withConnection(s"insert [${state.persistenceId}]") { connection =>
+              val session = new R2dbcSession(connection)
+              for {
+                updatedRows <- recoverDataIntegrityViolation(R2dbcExecutor.updateOneInTx(insertStatement(connection)))
+                _ <- handler.process(session, change)
```

Comment:
Would it be useful to decorate/wrap any exception here with details about that it was a change handler causing problems?